File: p567.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int len1 = s1.length();
        vector<int> cnt(26, 0);
        for (auto c : s1) {
            cnt[c-'a'] += 1;
        }
        
        int len2 = s2.length();
        vector<vector<int>> sum(26, vector<int>(len2+1, 0));
        for (int i = 0; i < len2; i++) {
            for (int j = 0; j < 26; j++) {
                sum[j][i+1] = sum[j][i];
            }
            sum[s2[i]-'a'][i+1] += 1;
        }
        
        bool suc = false;
        for (int i = 0; i <= len2-len1; i++) {
            bool ssuc = true;
            for (int j = 0; j < 26; j++) {
                ssuc &= (cnt[j] == sum[j][i+len1] - sum[j][i]);
            }
            suc |= ssuc;
        }
        return suc;
    }
};

File: python/2023_daily/p567.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        char_cnt = [0 for _ in range(26)]
        for c in s1:
            char_idx = ord(c) - ord("a")
            char_cnt[char_idx] += 1
        match_cnt = sum(1 if cnt == 0 else 0 for cnt in char_cnt)
        for idx in range(len(s1) - 1):
            # Update char_cnt and match_cnt
            char_idx = ord(s2[idx]) - ord("a")
            char_cnt[char_idx] -= 1
            if char_cnt[char_idx] == 0:
                match_cnt += 1
            elif char_cnt[char_idx] == -1:
                match_cnt -= 1
        for idx in range(len(s1) - 1, len(s2)):
            # Update char_cnt and match_cnt
            char_idx = ord(s2[idx]) - ord("a")
            char_cnt[char_idx] -= 1
            if char_cnt[char_idx] == 0:
                match_cnt += 1
            elif char_cnt[char_idx] == -1:
                match_cnt -= 1
            # Check if successful
            if match_cnt == 26:
                return True
            # Update char_cnt and match_cnt
            char_idx = ord(s2[idx - (len(s1) - 1)]) - ord("a")
            char_cnt[char_idx] += 1
            if char_cnt[char_idx] == 0:
                match_cnt += 1
            elif char_cnt[char_idx] == 1:
                match_cnt -= 1
        return False
